tests/benchmarks/bench_assembly.cpp

```diff
@@ -0,0 +1,389 @@
+   Problem<Kernel##Integrator,VDIM,p_eq_q> ker(AssemblyLevel::assembly, p, N);\
+   if ( !ker.is_runnable() ) { state.SkipWithError("MAX_MEM"); }\
+   while (state.KeepRunning()) { ker.benchmark(); }\
+   state.counters["MDof/s"] = bm::Counter(ker.SumMdofs(), bm::Counter::kIsRate);\
+   state.counters["Dofs"] = bm::Counter(ker.dofs, bm::Counter::kDefaults);\
+   state.counters["Order"] = bm::Counter(ker.p);\
+   state.counters["Assembly"] = (int)AssemblyLevel::assembly;}\
+BENCHMARK(BP##i##assembly)->ArgsProduct({\
+      benchmark::CreateRange(1024, max_dofs, /*step=*/2),\
+      benchmark::CreateDenseRange(1, max_order, /*step=*/1)\
+    })->Unit(bm::kMillisecond);
+
+// PARTIAL:
+/// BP1: scalar PCG with mass matrix, q=p+2
+BakeOff_Problem(PARTIAL,1,Mass,1,false)
+
+/// BP2: vector PCG with mass matrix, q=p+2
+BakeOff_Problem(PARTIAL,2,VectorMass,3,false)
+
+/// BP3: scalar PCG with stiffness matrix, q=p+2
+BakeOff_Problem(PARTIAL,3,Diffusion,1,false)
+
+/// BP4: vector PCG with stiffness matrix, q=p+2
+BakeOff_Problem(PARTIAL,4,VectorDiffusion,3,false)
+
+/// BP5: scalar PCG with stiffness matrix, q=p+1
+BakeOff_Problem(PARTIAL,5,Diffusion,1,true)
+
+/// BP6: vector PCG with stiffness matrix, q=p+1
+BakeOff_Problem(PARTIAL,6,VectorDiffusion,3,true)
+
+// ELEMENT:
+/// BP1: scalar PCG with mass matrix, q=p+2
+BakeOff_Problem(ELEMENT,1,Mass,1,false)
+
+/// BP2: vector PCG with mass matrix, q=p+2
+// BakeOff_Problem(ELEMENT,2,VectorMass,3,false)
+
+/// BP3: scalar PCG with stiffness matrix, q=p+2
+BakeOff_Problem(ELEMENT,3,Diffusion,1,false)
+
+/// BP4: vector PCG with stiffness matrix, q=p+2
+// BakeOff_Problem(ELEMENT,4,VectorDiffusion,3,false)
+
+/// BP5: scalar PCG with stiffness matrix, q=p+1
+BakeOff_Problem(ELEMENT,5,Diffusion,1,true)
+
+/// BP6: vector PCG with stiffness matrix, q=p+1
+// BakeOff_Problem(ELEMENT,6,VectorDiffusion,3,true)
+
+// FULL:
+/// BP1: scalar PCG with mass matrix, q=p+2
+BakeOff_Problem(FULL,1,Mass,1,false)
+
+/// BP2: vector PCG with mass matrix, q=p+2
+// BakeOff_Problem(FULL,2,VectorMass,3,false)
+
+/// BP3: scalar PCG with stiffness matrix, q=p+2
+BakeOff_Problem(FULL,3,Diffusion,1,false)
+
+/// BP4: vector PCG with stiffness matrix, q=p+2
+// BakeOff_Problem(FULL,4,VectorDiffusion,3,false)
+
+/// BP5: scalar PCG with stiffness matrix, q=p+1
+BakeOff_Problem(FULL,5,Diffusion,1,true)
+
+/// BP6: vector PCG with stiffness matrix, q=p+1
+// BakeOff_Problem(FULL,6,VectorDiffusion,3,true)
+
+/// Bake-off Kernels (BKs)
+template <typename BFI, int VDIM = 1, bool GLL = false>
+struct Kernel: public BakeOff
+{
+   GridFunction y;
+
+   Kernel(AssemblyLevel assembly, int order, int N)
+   : BakeOff(assembly,order,N,VDIM,GLL), y(&fes)
+   {
+      if (is_runnable())
+      {
+         x.Randomize(1);
+         a.SetAssemblyLevel(assembly);
+         a.AddDomainIntegrator(new BFI(one, GLL?irGLL:ir));
+         a.Assemble();
+         a.Mult(x, y);
+         MFEM_DEVICE_SYNC;
+      }
+   }
+
+   void setup() override
```

Comment:
Is this benchmark used?

Reply:
Not currently as it is redundant with the `Problem::setup`, do you have a proposition to reorganize this?